server/src/main/java/org/elasticsearch/ingest/IngestStats.java

```diff
@@ -234,52 +263,63 @@ static Stats merge(Stats first, Stats second) {
     static class Builder {
         private Stats totalStats = null;
         private final List<PipelineStat> pipelineStats = new ArrayList<>();
-        private final Map<String, List<ProcessorStat>> processorStats = new HashMap<>();
+        private final Map<ProjectId, Map<String, List<ProcessorStat>>> processorStats = new HashMap<>();
 
         Builder addTotalMetrics(IngestMetric totalMetric) {
             assert totalStats == null;
             this.totalStats = totalMetric.createStats();
             return this;
         }
 
-        Builder addPipelineMetrics(String pipelineId, IngestPipelineMetric ingestPipelineMetrics) {
+        Builder addPipelineMetrics(ProjectId projectId, String pipelineId, IngestPipelineMetric ingestPipelineMetrics) {
             this.pipelineStats.add(
-                new PipelineStat(pipelineId, ingestPipelineMetrics.createStats(), ingestPipelineMetrics.createByteStats())
+                new PipelineStat(projectId, pipelineId, ingestPipelineMetrics.createStats(), ingestPipelineMetrics.createByteStats())
             );
             return this;
         }
 
-        Builder addProcessorMetrics(String pipelineId, String processorName, String processorType, IngestMetric metric) {
-            this.processorStats.computeIfAbsent(pipelineId, k -> new ArrayList<>())
+        Builder addProcessorMetrics(
+            ProjectId projectId,
+            String pipelineId,
+            String processorName,
+            String processorType,
+            IngestMetric metric
+        ) {
+            this.processorStats.computeIfAbsent(projectId, k -> new HashMap<>())
+                .computeIfAbsent(pipelineId, k -> new ArrayList<>())
                 .add(new ProcessorStat(processorName, processorType, metric.createStats()));
             return this;
         }
 
         IngestStats build() {
-            return new IngestStats(totalStats, Collections.unmodifiableList(pipelineStats), Collections.unmodifiableMap(processorStats));
+            return new IngestStats(totalStats, unmodifiableList(pipelineStats), unmodifiableMap(processorStats));
         }
     }
 
     /**
      * Container for pipeline stats.
      */
-    public record PipelineStat(String pipelineId, Stats stats, ByteStats byteStats) {
+    public record PipelineStat(ProjectId projectId, String pipelineId, Stats stats, ByteStats byteStats) {
         static List<PipelineStat> merge(List<PipelineStat> first, List<PipelineStat> second) {
-            var totalsPerPipeline = new HashMap<String, PipelineStat>();
+            record MergeKey(ProjectId projectId, String pipelineId) {}
```

Comment:
TIL you could define records inside methods! Do you have any idea how that compiles? I'm mostly curious if the compiler extracts that record definition somewhere or whether it's going to redefine that record on every invocation of this method.

Reply:
No I think this is perfectly fine. Thanks for the explanation!